File: public/java/037-serialize-deserialize-binary-tree.java
/*
 * LeetCode Problem #297: Serialize and Deserialize Binary Tree
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/
 */
import java.util.*;

public class Solution {

    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // ─── APPROACH 1: Pre-order DFS ───────────────────────────────────────────
    // APPROACH 1: Pre-order DFS | O(n) time | O(n) space
    // EXPLAIN: Serialize with pre-order DFS, marking nulls as "null"; deserialize by consuming a string-stream token by token.
    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        serHelper(root, sb);
        return sb.toString();
    }
    private void serHelper(TreeNode node, StringBuilder sb) {
        if (node == null) { sb.append("null,"); return; }
        sb.append(node.val).append(',');
        serHelper(node.left,  sb);
        serHelper(node.right, sb);
    }
    public TreeNode deserialize(String data) {
        Queue<String> q = new LinkedList<>(Arrays.asList(data.split(",")));
        return desHelper(q);
    }
    private TreeNode desHelper(Queue<String> q) {
        String val = q.poll();
        if ("null".equals(val)) return null;
        TreeNode node = new TreeNode(Integer.parseInt(val));
        node.left  = desHelper(q);
        node.right = desHelper(q);
        return node;
    }

    // APPROACH 2: Level-order BFS | O(n) time | O(n) space
    // EXPLAIN: BFS serializes nodes level by level; deserialization reconstructs by assigning children to each dequeued node.
    public String serializeBFS(TreeNode root) {
        if (root == null) return "";
        StringBuilder sb = new StringBuilder();
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            if (node == null) { sb.append("null,"); continue; }
            sb.append(node.val).append(',');
            q.offer(node.left);
            q.offer(node.right);
        }
        return sb.toString();
    }
    public TreeNode deserializeBFS(String data) {
        if (data.isEmpty()) return null;
        String[] tokens = data.split(",");
        TreeNode root = new TreeNode(Integer.parseInt(tokens[0]));
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int i = 1;
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            if (i < tokens.length && !tokens[i].equals("null")) {
                node.left = new TreeNode(Integer.parseInt(tokens[i])); q.offer(node.left);
            }
            i++;
            if (i < tokens.length && !tokens[i].equals("null")) {
                node.right = new TreeNode(Integer.parseInt(tokens[i])); q.offer(node.right);
            }
            i++;
        }
        return root;
    }

    // APPROACH 3: Pre-order with Space Separator | O(n) time | O(n) space
    // EXPLAIN: Variation of Approach 1 using '#' for null and space as delimiter; demonstrates flexible encoding choices.
    public String serializeSpace(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        spaceHelper(root, sb);
        return sb.toString().trim();
    }
    private void spaceHelper(TreeNode node, StringBuilder sb) {
        if (node == null) { sb.append("# "); return; }
        sb.append(node.val).append(' ');
        spaceHelper(node.left,  sb);
        spaceHelper(node.right, sb);
    }
    public TreeNode deserializeSpace(String data) {
        Queue<String> q = new LinkedList<>(Arrays.asList(data.split("\\s+")));
        return spaceDeHelper(q);
    }
    private TreeNode spaceDeHelper(Queue<String> q) {
        String val = q.poll();
        if (val == null || "#".equals(val)) return null;
        TreeNode node = new TreeNode(Integer.parseInt(val));
        node.left  = spaceDeHelper(q);
        node.right = spaceDeHelper(q);
        return node;
    }

    // APPROACH 4: Compact Parentheses | O(n) time | O(n) space
    // EXPLAIN: Encode as "val(left)(right)"; omit right-branch parens when right is null; parse with a mutable index.
    public String serializeCompact(TreeNode root) {
        if (root == null) return "";
        String result = String.valueOf(root.val);
        if (root.left != null || root.right != null)
            result += "(" + serializeCompact(root.left) + ")";
        if (root.right != null)
            result += "(" + serializeCompact(root.right) + ")";
        return result;
    }
    private int[] idx = {0};
    public TreeNode deserializeCompact(String data) {
        idx[0] = 0;
        return compactHelper(data);
    }
    private TreeNode compactHelper(String data) {
        if (idx[0] >= data.length()) return null;
        int start = idx[0];
        if (data.charAt(idx[0]) == '-') idx[0]++;
        while (idx[0] < data.length() && Character.isDigit(data.charAt(idx[0]))) idx[0]++;
        if (start == idx[0]) return null;
        TreeNode node = new TreeNode(Integer.parseInt(data.substring(start, idx[0])));
        if (idx[0] < data.length() && data.charAt(idx[0]) == '(') {
            idx[0]++; node.left  = compactHelper(data); idx[0]++;
        }
        if (idx[0] < data.length() && data.charAt(idx[0]) == '(') {
            idx[0]++; node.right = compactHelper(data); idx[0]++;
        }
        return node;
    }
}

// Made with Bob

File: public/python/037-serialize-deserialize-binary-tree.py
# LeetCode Problem #297: Serialize and Deserialize Binary Tree
# Difficulty: Hard
# Link: https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

from typing import Optional
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# ─────────────────────────────────────────────
# APPROACH 1: Pre-order Traversal (DFS) | O(n) time | O(n) space
# EXPLAIN: Serialize using pre-order DFS with 'null' markers; deserialize by consuming tokens left-to-right recursively.
# WHEN: Most common interview approach; pre-order naturally rebuilds trees top-down.

class Codec_Preorder:
    def serialize(self, root: Optional[TreeNode]) -> str:
        parts = []
        def dfs(node):
            if not node:
                parts.append('null')
                return
            parts.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        return ','.join(parts)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        tokens = iter(data.split(','))
        def build():
            val = next(tokens)
            if val == 'null':
                return None
            node = TreeNode(int(val))
            node.left  = build()
            node.right = build()
            return node
        return build()


# ─────────────────────────────────────────────
# APPROACH 2: Level-order Traversal (BFS) | O(n) time | O(n) space
# EXPLAIN: BFS serializes level by level; deserialization rebuilds by assigning children to each dequeued node.
# WHEN: More intuitive for those who think in levels; matches the typical array representation.

class Codec_BFS:
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ''
        parts = []
        q = deque([root])
        while q:
            node = q.popleft()
            if node:
                parts.append(str(node.val))
                q.append(node.left)
                q.append(node.right)
            else:
                parts.append('null')
        return ','.join(parts)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        if not data:
            return None
        tokens = data.split(',')
        root = TreeNode(int(tokens[0]))
        q = deque([root])
        i = 1
        while q:
            node = q.popleft()
            if i < len(tokens) and tokens[i] != 'null':
                node.left = TreeNode(int(tokens[i]))
                q.append(node.left)
            i += 1
            if i < len(tokens) and tokens[i] != 'null':
                node.right = TreeNode(int(tokens[i]))
                q.append(node.right)
            i += 1
        return root


# ─────────────────────────────────────────────
# APPROACH 3: Pre-order with Space Separator | O(n) time | O(n) space
# EXPLAIN: Same as Approach 1 but uses '#' for null and space as delimiter; parsed with a string stream.
# WHEN: Variation of Approach 1 showing different delimiter/marker choices.

class Codec_Space:
    def serialize(self, root: Optional[TreeNode]) -> str:
        parts = []
        def dfs(node):
            if not node:
                parts.append('#')
                return
            parts.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        return ' '.join(parts)

    def deserialize(self, data: str) -> Optional[TreeNode]:
        tokens = iter(data.split())
        def build():
            val = next(tokens, None)
            if val is None or val == '#':
                return None
            node = TreeNode(int(val))
            node.left  = build()
            node.right = build()
            return node
        return build()


# ─────────────────────────────────────────────
# APPROACH 4: Compact Parentheses Representation | O(n) time | O(n) space
# EXPLAIN: Encode as "val(left)(right)" — omit branches that are empty leaves; parse recursively.
# WHEN: More space-efficient string; shows creativity in encoding design.

class Codec_Compact:
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ''
        result = str(root.val)
        if root.left or root.right:
            result += '(' + self.serialize(root.left) + ')'
        if root.right:
            result += '(' + self.serialize(root.right) + ')'
        return result

    def deserialize(self, data: str) -> Optional[TreeNode]:
        if not data:
            return None
        self._pos = [0]
        return self._build(data)

    def _build(self, data: str) -> Optional[TreeNode]:
        pos = self._pos
        if pos[0] >= len(data):
            return None
        start = pos[0]
        if data[pos[0]] == '-':
            pos[0] += 1
        while pos[0] < len(data) and data[pos[0]].isdigit():
            pos[0] += 1
        if start == pos[0]:
            return None
        node = TreeNode(int(data[start:pos[0]]))
        if pos[0] < len(data) and data[pos[0]] == '(':
            pos[0] += 1  # skip '('
            node.left = self._build(data)
            pos[0] += 1  # skip ')'
        if pos[0] < len(data) and data[pos[0]] == '(':
            pos[0] += 1
            node.right = self._build(data)
            pos[0] += 1
        return node


# Made with Bob
